### PR TITLE
Sdma fixes take3

### PR DESCRIPTION
**Pull request by dbaluta**
This is the last series of SDMA fixes. With this playback works fine and the sound is clear!

### CODE REVIEW

File: src/drivers/imx/sdma.c [hunk 1]
```diff
@@ -159,6 +159,8 @@ static int sdma_register_init(struct dma *dma)
 		return ret;
 	}
 
+	dma_reg_write(dma, SDMA_MC0PTR, 0);
+
```
Comment (paulstelian97) [inline src/drivers/imx/sdma.c:164]: Wait, when is it written?

File: src/include/sof/drivers/sdma.h [hunk 2]
```diff
@@ -78,6 +78,7 @@
 #define SDMA_DEFPRI		4
 #define SDMA_MAXPRI		7
 
+#define SDMA_CHANNELS_COUNT	32
```
Comment (paulstelian97) [inline src/include/sof/drivers/sdma.h:81]: Why don't we use the .count field?

File: src/drivers/imx/sdma.c [hunk 3]
```diff
@@ -579,23 +580,31 @@ static int sdma_release(struct dma_chan_data *channel)
...
 	for (int i = 0; i < pdata->desc_count; i++) {
 		/* Work around the fact that we cannot allocate uncached memory
 		 * on all platforms supporting SDMA.
 		 */
 		dcache_invalidate_region(&pdata->desc[i].config,
 					 sizeof(pdata->desc[i].config));
-		pdata->desc[i].config |= SDMA_BD_DONE;
+		if (i == pdata->next_bd)
+			pdata->desc[i].config |= SDMA_BD_DONE;
+
```
Comment (paulstelian97) [inline src/drivers/imx/sdma.c:601]: Can't we get rid of the loop overall?

File: src/drivers/imx/sdma.c (hunk 1 above)
Comment (dbaluta) [reply src/drivers/imx/sdma.c:164]: @paulstelian97 it is written later in the same function
```
»       dma_reg_write(dma, SDMA_MC0PTR, (uint32_t)pdata->ccb_array);
```
Also, the reset value should already be 0 but I've seen Linux driver doing this so for consistency I am also doing it here.

File: src/include/sof/drivers/sdma.h (hunk 2 above)
Comment (dbaluta) [reply src/include/sof/drivers/sdma.h:81]: we could, but this is more like MAX HW channels and .count field says how many channels we actually want to use.

Comment (paulstelian97) [reply src/include/sof/drivers/sdma.h:81]: Are we confident that all SDMA implementations will have 32 channels in the future?

File: src/drivers/imx/sdma.c (hunk 3 above)
Comment (dbaluta) [reply src/drivers/imx/sdma.c:601]: Yes, this can be done use directly pdata->next_bd as the index of desc to be modified.

Comment (dbaluta) [reply src/drivers/imx/sdma.c:604]: @paulstelian97 OK, fixed!

File: src/include/sof/drivers/sdma.h (hunk 2 above)
Comment (paulstelian97) [reply src/include/sof/drivers/sdma.h:81]: It's safer to use dma->plat_data.channels. I don't think we'll put it to something else anyway -- we could probably create another limit in channel_get per platform.

This driver works standalone on the full device and thus should need to know how many channels are in-hardware anyway. We don't have a driver that works as a slave yet, with the actual allocation of channels and managing of them done on the ARM side.

Comment (dbaluta) [reply src/include/sof/drivers/sdma.h:81]: ok, no problem. will fix that.

**Pull request merged**